docs: 3.0 migration guide (#2606)
XRPLF/xrpl.js: MIGRATION.md

# Migration Guide

> The goal of xrpl.js 3.0 was to improve the libraries support for non nodejs projects by eliminating the need for platform/build system specific polyfills and se.  This was accomplished 
> 
> There were five main goals for xrpl.js 3.0 package suite:
> 
> - Remove polyfills as dependencies for non node.js projects
> - Reduce the bundle size for client side applications
> - Execute unit tests in a browser environment
> - Few breaking changes
> - Improve docs
> 
> The result of which is a 60% reduction in bundle size and a complete elimination of polyfills for browser applications as well as a streamlining of react-native support.  All sub packages now run their tests in both nodejs and Chrome. The `xrpl` still will not run its unit tests in the browser but it still has exhaustive coverage through its integration tests which do.

Reviewer: Proposed rewrite. Reasoning:
* It's good to lead with reason to migrate (in this case massive bundle size reduction and elimination of polyfills)
* It's important to telegraph the big changes that they'll need to pay attention to (which is primarily the UInt8Array -> Buffer changes)
* The improvement in unit testing, while important from a reliability standpoint, doesn't impact their migration, so should be moved to a more informational section.

```suggestion
Xrpl.js 3.0 reduces bundle size by 60% and completely removes the need for polyfills for browser applications. It does this with minimal breaking changes. It accomplished this by replacing node specific dependencies with ones which would work in the browser.

The primary breaking change is updating all `Buffer` types to `UInt8Array` since `Buffer` is not supported in the browser. `Buffer` objects can usually be directly replaced with `UInt8Array` as `Buffer` extends `UInt8Array`. The primary difference in functionality is that `Buffer` has additional helper functions. We've listed all functions affected by this change below in the `UInt8Array` section.
```

> ## Polyfill Reduction

Reviewer: You did more than reduce :)
```suggestion
## Eliminating Polyfills
```

The polyfill elimination was achieved by switching from elliptic, create-hash, and crypto polyfills to the `@noble` suite of packages which includes `@noble/hashes`, `@noble/curves`, `@scure/bip32`, @scure/bip39, and `@scure/base`.  In order to still use node's built in crypto functions when running in a node environment `@xrplf/isomorphic` was created as a way to abstract the selection of the underlying implementation.

Additionally the faucet code switched from using node's `http` packages and instead leverage native fetch in the browser and `node-fetch` in node. This eliminated the polyfills for `http`, `https`, and `url`. The need for `assert` was addressed by switching to conditionals that throw exceptions and `utils` by updating error handling to use `JSON.stringify` instead of `inspect`.

> The removal of `buffer` proved to be a tricky endeavor.  It brought along with it the largest amount of breaking changes but was necessary to get to a polyfill free future.  Since the `Buffer` object is not native to the browser all apis were migrated to `Uint8Arrays`.  Many libraries have begun making this transition as well so with the goal in sight it made since to address it at this time. Checkout this [blogpost](https://sindresorhus.com/blog/goodbye-nodejs-buffer) by Sindre Sorhus on the rational behind using `UintArrays` over `Buffer`.

Reviewer: nit:
```suggestion
The removal of `buffer` proved to be a tricky endeavor.  It brought along with it the largest amount of breaking changes but was necessary to get to a polyfill free future.  Since the `Buffer` object is not native to the browser all apis were migrated to `Uint8Arrays`.  Many libraries have begun making this transition as well so with the goal in sight it made since to address it at this time. Checkout this [blogpost](https://sindresorhus.com/blog/goodbye-nodejs-buffer) by Sindre Sorhus on the rationale behind using `UintArrays` over `Buffer`.
```

- `Quality.decode` returns a `BigNumber` instead of a `Decimal`
- `Amount.assertIouIsValid` take a `BigNumber` instead `Decimal`
- `Amount.verifyNoDecimal` takes a `BigNumber` instead `Decimal`

### Error Handling

In order to remove `assert` as a dependency regular `Error` objects are thrown instead of  `AssertionError`.

Not a breaking change but there is now a much more descriptive error message when keys do not match for a given algorithm.  This impacts most of `ripple-keypairs` functions but only if you already had issues with incompatible values.

### Bundlers

For `vite` and `create-react-app` you can remove all xrpl.js polyfills/configurations this also includes the custom mappings for `ws` to `WsWrapper` and the exclusion of `https-proxy-agent`. These improves should extend to using the libraries with other bundlers or frameworks but has only been tested in those noted above.

#### React Native

Neither `react-native-crypto` nor `rn-nodeify` are required.  Instead `fast-text-encoding` and `react-native-get-random-values` are as well as a small polyfills file to wire them up. You can find the new guide at [UNIQUE_SETUPS.md](./UNIQUE_SETUPS.md).

### `xrpl-secret-numbers`

> This library was absorbed into the monorepo as it was causing dependency mismatches and was going to require an update anyway.  It is now available as `@xrplf/secret-numbers`.  Besides making changes to use `UInt8Array` you will need to update all places where you use functions on the `Util` object.  These methods are now at the root of the library. These methods include:

Reviewer: Is the `Util` object specific to `xrpl-secret-numbers` or should this be it's own section?

- `Utils.randomEntropy`,
- `Utils.randomSecret`
- `Utils.entropyToSecret`
- `Utils.secretToEntropy`
- `Utils.calculateChecksum`
- `Utils.checkChecksum`
- `Utils.parseSecretString`

##### Prebundle

If you use the pre bundled version of the library you will need to make the following changes:

- Change any references to `dist/browerified.js` to `build/xrplf-secret-numbers-latest.js`.
- Access any methods as properties of `xrplf_secret_numbers` instead of using browserify's loader.

### HttpsProxyAgent

> The way to configure proxies for `Client` has changed. It is now done by specifying the `agent` parameter on the ConnectionOptions config. This can use be created by libraries such as `https-proxy-agent` or any that

Reviewer: nit:
```suggestion
The way to configure proxies for `Client` has changed. It is now done by specifying the `agent` parameter on the `ConnectionOptions` config. This can use be created by libraries such as `https-proxy-agent` or any that
```

implements `http.Agent`.  This was done to remove a hard dependency on `https-proxy-agent` when running in the browser. The lack of proxy support is not a reduction in features as it never supported running in the browser and merely burdened bundles with unused dependencies.  This was also done to support `https-proxy-agent@7` which changed several option names.

#### Before

```javascript
{
  proxy: `ws://127.0.0.1:${port}`,
  authorization: 'authorization',
  trustedCertificates: ['path/to/pem'],
}
```

#### After

```javascript
 {
   agent: new HttpsProxyAgent<string>(`ws://127.0.0.1:${port}`, {
     ca: ['path/to/pem'],
   }),
   authorization: 'authorization'

>  }

Reviewer: Please include the context in which this would've been used previously (e.g. one line with the Client object having that parameter set).

```

### Transaction` type

> The type `Transaction` exported by `xrpl` was renamed to `SubmittableTransaction` and the definition for `Transaction` was then changed to include all transactions which is a union of `SubmittableTransaction` and `PseudoTransaction`.  This effectively changes the signature of the following methods:

Reviewer: The `Transaction` type has been extended to include `PsuedoTransaction`s. The previous subset of transactions was just those which were submittable, so they are now represented by the type `SubmittableTransaction`. This changes the signature of the following methods:

> The type `Transaction` exported by `xrpl` was renamed to `SubmittableTransaction` and the definition for `Transaction` was then changed to include all transactions which is a union of `SubmittableTransaction` and `PseudoTransaction`.  This effectively changes the signature of the following methods:

Reviewer: ```suggestion
The `Transaction` type has been extended to include `PsuedoTransaction`'s. The previous subset of transactions was just those which were submittable, so they are now represented by the type `SubmittableTransaction`. This changes the signature of the following methods:
```

# Migration Guide

> In Xrpl.js 3.0, we've made significant improvements that result in a 60% reduction in bundle size for browser applications. We've also eliminated the need for polyfills with minimal disruption to existing code. This was achieved by replacing node-specific dependencies with ones that are compatible with browsers.

Reviewer 2: ```suggestion
In xrpl.js 3.0, we've made significant improvements that result in a 60% reduction in bundle size for browser applications. We've also eliminated the need for polyfills with minimal disruption to existing code. This was achieved by replacing Node specific dependencies with ones that are compatible with browsers.
```

The main change you'll notice is the update replacing `Buffer` with `Uint8Array` across the board. This was done since browsers don't support `Buffer`. Fortunately, this transition is relatively straightforward, as `Buffer` is a subclass of `Uint8Array`, meaning in many circumstances `Buffer` can be directly replaced by `Uint8Array`. The primary difference is that `Buffer` has additional helper functions. We've listed the affected client library functions below in the `Uint8Array` section for your reference.

This migration guide also applies to:
`ripple-address-codec` 4.3.1 -> 5.0.0
`ripple-binary-codec` 1.11.0 -> 2.0.0
`ripple-keypairs` 1.3.1 -> 2.0.0

# Why update to 3.0?

At a high level:

1. 60% filesize reduction
2. Simplified setup by removing polyfills
3. Increased reliability through more browser testing

<aside>
💡 Also, with 3.0, the [reference docs](https://js.xrpl.org/) have received a massive update to include examples, more readable type information, and properly display documentation on functions which are part of the `Client` class.
</aside>

## 1. 60% size reduction

> Through simplifying 3rd party dependencies, xrpl.js is now a much ligher-weight library.

Reviewer 2: ```suggestion
Through simplifying 3rd party dependencies, xrpl.js is now a much lighter-weight library.
```

# Migration Guide

> Xrpl.js 3.0 reduces bundle size by 60% and completely removes the need for polyfills for browser applications. It does this with minimal breaking changes. It accomplished this by replacing node specific dependencies with ones which would work in the browser.
> 
> The primary breaking change is updating all `Buffer` types to `Uint8Array` since `Buffer` is not supported in the browser. `Buffer` objects can usually be directly replaced with `UInt8Array` as `Buffer` extends `Uint8Array`. The primary difference in functionality is that `Buffer` has additional helper functions. We've listed all functions affected by this change below in the `Uint8Array` section.

Reviewer 3: In Xrpl.js 3.0, we've made significant improvements that result in a 60% reduction in bundle size for browser applications. We've also eliminated the need for polyfills with minimal disruption to existing code. This was achieved by replacing node-specific dependencies with ones that are compatible with browsers.

The main change you'll notice is the update of all Buffer types to Uint8Array because browsers don't support Buffer. Fortunately, this transition is relatively straightforward, as Buffer can often be replaced directly with UInt8Array, as Buffer extends Uint8Array. The main functional difference is that Buffer has additional helper functions, which we've listed in the Uint8Array section below for your reference.
```suggestion
Xrpl.js 3.0 reduces bundle size by 60% and completely removes the need for polyfills for browser applications. It does this with minimal breaking changes. It accomplished this by replacing node specific dependencies with ones which would work in the browser.

The primary breaking change is updating all `Buffer` types to `Uint8Array` since `Buffer` is not supported in the browser. `Buffer` objects can usually be directly replaced with `UInt8Array` as `Buffer` extends `Uint8Array`. The primary difference in functionality is that `Buffer` has additional helper functions. We've listed all functions affected by this change below in the `Uint8Array` section.
```

This migration guide also applies to:
`ripple-address-codec` 4.3.1 -> 5.0.0
`ripple-binary-codec` 1.11.0 -> 2.0.0
`ripple-keypairs` 1.3.1 -> 2.0.0

# Why update to 3.0?

At a high level:

1. 60% filesize reduction
2. Simplified setup by removing polyfills

> 3. Increased reliability

Reviewer 3: Is there a way we could quickly list how reliability is improved? In the previous 2 bullets you said the how, but this bullet is just a statement with no support

<aside>
💡 Also, with 3.0, the [reference docs](https://js.xrpl.org/) have received a massive update to include examples, more readable type information, and properly display documentation on functions which are part of the `Client` class.
</aside>

> ## 1. 60% size reduction
> 
> Through simplifying 3rd party dependencies, xrpl.js is now a much ligher-weight library.
> 
> One major source of the project’s large bundle size was polyfills, code that replicated Node specific features in the browser. Instead, we shifted to using 3rd party packages which worked in both Node and the browser by default.
> 
> A simple fix was removing `lodash` by using es6 array methods and porting over simple helper utilities.
> 
> The last main source of the reduction was simplifying the 5 large number libraries down to one.
> 
> Before, we had `decimal.js`, `bit-integer`, `bitnumber.js`, and two versions of `bn.js` due to `elliptic`'s transitive dependency tree.
> 
> After, we used `@noble` to replace `elliptic` to have just one version of `bn.js`, and normalized our library to just use `bignumber.js` across the board.

Reviewer 3: ## 1. 60% Size Reduction

We've significantly reduced the size of xrpl.js by 60%, making it a much lighter-weight library.

A major contributor to the project's large bundle size was the use of polyfills, which replicated Node-specific features in the browser. To address this, we transitioned to utilizing 3rd party packages that inherently supported both Node and browser environments.

One straightforward optimization involved eliminating the use of `lodash` by employing ES6 array methods and migrating simple helper utilities.

Another substantial reduction came from simplifying the five large number libraries down to just one. Previously, we relied on `decimal.js`, `bit-integer`, `bitnumber.js`, and two versions of `bn.js` due to `elliptic`'s transitive dependency tree. However, we streamlined this by adopting `@noble` to replace `elliptic`, resulting in the use of just one version of `bn.js`. This allowed us to standardize our library to use `bignumber.js` consistently across the board.
```suggestion
## 1. 60% size reduction

Through simplifying 3rd party dependencies, xrpl.js is now a much ligher-weight library.

One major source of the project’s large bundle size was polyfills, code that replicated Node specific features in the browser. Instead, we shifted to using 3rd party packages which worked in both Node and the browser by default.

A simple fix was removing `lodash` by using es6 array methods and porting over simple helper utilities.

The last main source of the reduction was simplifying the 5 large number libraries down to one.

Before, we had `decimal.js`, `bit-integer`, `bitnumber.js`, and two versions of `bn.js` due to `elliptic`'s transitive dependency tree.

After, we used `@noble` to replace `elliptic` to have just one version of `bn.js`, and normalized our library to just use `bignumber.js` across the board.
```

## 2. No more polyfills required (simplified install)

> Polyfills made it hard to setup xrpl.js in the browser as they required custom bundler configs. By using dependencies and browser-native features, xrpl.js can now work just by installing from `npm` in most cases.

Reviewer 3: Is there any callout we can make about what cases it wouldn't work in?

Reviewer: Yep, specifically `react-native` still needs them.

For the cryptography libraries, we switched from using `elliptic`, `create-hash`, and other crypto polyfills to using the `@noble` suite of packages. For situations where node-specific crypto functions performed better, we created `@xrplf/isomorphic` to dynamically choose which implementation to use depending on the runtime environment.

<aside>
💡 The `@noble` suite of packages includes `@noble/hashes`, `@noble/curves`, `@scure/bip32`, `@scure/bip39`, and `@scure/base`

> 

Reviewer 3: Can we remove this space?

</aside>

We eliminated the polyfills for `http`, `https`, and `url` by using the native `fetch` in the browser.

The easiest to replace were `assert` (which was replaced by simple conditions & exceptions) and `utils` (which used `JSON.stringify` instead of `inspect`).

> Lastly, the `buffer` polyfill turned out to be the trickiest to remove, resulting in the largest number of breaking changes. Since the `Buffer` object is not native to the browser all apis were migrated to the superclass of `Buffer` → `Uint8Arrays`. For a detailed write up of why we and many libraries are choosing to make this transition, check out this [blog post](https://sindresorhus.com/blog/goodbye-nodejs-buffer) by Sindre Sorhus.

Reviewer 3: I think it should be `Uint8Array`s instead of `Uint8Arrays`

## 3. Increased Reliability

> With xrpl.js 3.0, we improved our test coverage in the browser. Specifically, we added browser unit testing to all packages in the monorepo other than the `xrpl` package. Note that the `xrpl` package has browser coverage through our integration tests.
> 
> We achieved this by adding a `karma` config to every project that uses webpack to bundle each lib, then run the tests in Chrome. Hopefully soon we’ll also be able to add support for the `xrpl` package to run it’s unit tests in Chrome as part of our CI process.

Reviewer 3: In Xrpl.js 3.0, we enhanced our testing capabilities for browser compatibility. This improvement involved the introduction of browser unit testing for all packages within the monorepo, except for the xrpl package, which already had browser coverage through integration tests.
```suggestion
With xrpl.js 3.0, we improved our test coverage in the browser. Specifically, we added browser unit testing to all packages in the monorepo other than the `xrpl` package. Note that the `xrpl` package has browser coverage through our integration tests.

We achieved this by adding a `karma` config to every project that uses webpack to bundle each lib, then run the tests in Chrome. Hopefully soon we’ll also be able to add support for the `xrpl` package to run it’s unit tests in Chrome as part of our CI process.
```

To implement this enhancement, we included a karma configuration in every project utilizing webpack to bundle each library, allowing us to execute tests in the Chrome browser. We are actively working towards extending this support to include running unit tests for the xrpl package in Chrome as an integral part of our continuous integration (CI) process in the near future.

# Breaking Changes Detailed Migration Guide

Here’s a high-level overview of the breaking changes.

<aside>
💡 Note that the vast majority of these changes are very small typing changes, which should have direct 1-line replacements.

> 

Reviewer 3: Remove space?

</aside>

1. The largest change is that all instances of `Buffer` have been replaced by `Uint8Array` **[Link](#1-buffer-to-uint8array)**
2. All “large number” types have been consolidated to either `bigint` or `BigNumber` **[Link](#2-large-number-handling)**
3. Polyfill configuration changes **[Link](#3-polyfill-configuration-changes)**
4. `dropsToXRP` and `Client.getXrpBalance` now return a `number` instead of a `string` **[Link](#4-dropstoxrp-and-clientgetxrpbalance-now-return-a-number-instead-of-a-string)**
5. `xrpl-secret-numbers` has been moved into the mono-repo as `@xrplf/secret-numbers`  **[Link](#5-xrpl-secret-numbers-has-been-moved-into-the-mono-repo-as-xrplfsecret-numbers)**
6. Support for Node 14 has been dropped **[Link](#6-support-for-node-14-has-been-dropped)**
7. Configuring proxies with the Client  **[Link](#7-configuring-proxies-with-the-client)**
8. Bug fix: Setting an explicit `algorithm` when generating a wallet works now **[Link](#8-bug-fix-setting-an-explicit-algorithm-when-generating-a-wallet-works-now)**
9. `AssertionError` → `Error` **[Link](#9-assertionerror-→-error)**
10. Pre-bundle browser builds **[Link](#10-pre-bundle-browser-builds)**

> 11. We’ve updated the `Transaction` type to include psuedotransactions **[Link](#11-transaction-type)**

Reviewer 4: typo in spelling of the word `pseudo-transactions`

 }`

### 8. Bug fix: Setting an explicit `algorithm` when generating a wallet works now

`Wallet.generate()` and `Wallet.fromSeed` were ignoring the `algorithm` parameter. This means that if you were manually specifying `algorithm` in any `Wallet` constructors, you may generate a different `Wallet` keypair when upgrading to 3.0. In that case to get the same generated wallets as before, don’t specify the `algorithm` parameter.

### 9. `AssertionError` → `Error`

In order to get rid of the `assert` polyfill, we’ve replaced `AssertionError`s with `Error` exceptions. We’ve also updated the error messages to be more descriptive. If you were catching those specific errors, you will have to update your catch statements.

This impacts most of `ripple-keypairs` functions but only if you already had issues with incompatible values.

### 10. Pre-bundle browser builds

If you use the pre bundled version of the library you will need to make the following changes:

- Change any references to `dist/browerified.js` to `build/xrplf-secret-numbers-latest.js`.
- Access any methods as properties of `xrplf_secret_numbers` instead of using browserify's loader.

### 11. Transaction` type

> `Transaction` has been updated to include `PsuedoTransaction`s. To get the equivalent of the old `Transaction` type which only included transactions users could submit, please use `SubmittableTransaction`.

Reviewer 4: typo in `PseudoTransaction`. The `*.ts` code files have the correct spelling of the word

This effectively changes the signature of the following methods:

- `Client.autofill`
- `Client.submit`
- `Client.submitAndWait`
- `Client.prepareTransaction`
- `getSignedTx`
- `isAccountDelete`

### 12. `authorizeChannel` was moved

We’ve moved `authorizeChannel` from `wallet/signer` to `wallet/authorizeChannel` to solve a circular dependency issue. You may have to update your import path as a result.

### 13. We’ve removed the deprecated `BroadcastClient`

This feature was never fully implemented, and was marked as deprecated for several years. With 3.0 we’ve fully removed any code relating to it.

> # Wrap up

Reviewer 3: We should probably make mention of the minimum browser versions supported across desktop and mobile

Reviewer: Ended up not being able to find this in a reasonable amount of time - if we find this later, happy to add it as a separate small PR.
